pkg/statsd/dispatcher.go

```diff
@@ -51,28 +52,30 @@ func NewMetricDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorF
 }
 
 // Run runs the MetricDispatcher.
-func (d *MetricDispatcher) Run(ctx context.Context, done gostatsd.Done) {
-	defer done()
-	var wg sync.WaitGroup
-	wg.Add(d.numWorkers)
-	for _, worker := range d.workers {
-		w := worker // Make a copy of the loop variable! https://github.com/golang/go/wiki/CommonMistakes
-		go w.work(wg.Done)
-	}
+func (d *MetricDispatcher) Run(ctx context.Context) {
+	var wg wait.Group
 	defer func() {
 		for _, worker := range d.workers {
 			close(worker.metricsQueue) // Close channel to terminate worker
 		}
 		wg.Wait() // Wait for all workers to finish
 	}()
+	for _, worker := range d.workers {
+		wg.Start(worker.work)
+	}
 
 	// Work until asked to stop
 	<-ctx.Done()
 }
 
-func (d *MetricDispatcher) runMetrics(ctx context.Context, statser statser.Statser) {
+func (d *MetricDispatcher) RunMetrics(ctx context.Context, statser statser.Statser) {
```

Comment:
I would say this is an application that exists as a cohesive whole, not a library to pick and choose bits from to use elsewhere.